classes/Search.php

```diff
@@ -319,21 +319,18 @@ public static function find(
                 }
             }
 
-            if (!$result) {
+            if (empty($result)) {
                 unset($words[$key]);
                 continue;
             }
 
             $productIds = array_column($result, 'id_product');
-            if ($eligibleProducts2 === null) {
-                $eligibleProducts2 = $productIds;
-            } else {
-                $eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);
-            }
-
+            $eligibleProducts2 = array_merge($eligibleProducts2, $productIds);
```

Comment:
This is a big change of behavior.
Previously we would look at the intersection of `$eligibleProducts2` and `$productIds` now we merge them. I expect that this will make the search less accurate, it will present way much results than before.

I am not sure this is the right way to fix the issue

> It's this line: $eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);
This will return the differences between the two arrays which is none when you use 2 or more search words and so the result is empty.

Can you tell us
- what is inside `$eligibleProducts2` when a single search word is used
- what is inside `$eligibleProducts2` when two search words are used

I think the root issue might be there.

Reply:
@matks : If the two arrays don't match.....the array $eligibleProducts2 is empty. So if word 1 result is (e.g.) id_product 5 and word 2 result is id_product 6. You have NO results because they don't match. That's wrong in every way you look at it.

What is inside $eligibleProducts2 when a single search word is used:
Example: Search for testproduct1; the contents of $eligibleProducts2 = array([0]=>1)

What is inside $eligibleProducts2 when two search words are used:
Example: Search for testproduct1;testproduct2; the contents of $eligibleProducts2 = array()
Reason: The 2 arrays don't match and so array_intersect returns an empty array

What is inside $eligibleProducts2 when three search words are used:
Example: Search for testproduct1;testproduct2;testproduct3; the contents of $eligibleProducts2 = array()
Reason: The 2 arrays don't match and so array_intersect returns an empty array

Reply:
Aaaah I got it now.

I think that the goal of `$eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);` was to make search more accurate.

Exemple: I search "red shirt".
First loop on "red" will give me 4 red products
Second loop on "shirt" will give me 5 shirt products
Then intersection will give me the red shirt products

If we remove the `array_intersect()` we will get every red product and every shirt product. We don't want that to happen.

But if you search, like you do, the words "shirt ; trousers" you want to find both shirts and trousers
but the above logic will find all products that are shirts AND trousers and this is not possible.

So the main issue is that the current logic attempts to find "products that match all input words" while in your usecase you want "products that match any input word".

So I cannot approve your change. We need to find a way that preserve both behavior, if that's possible.

Reply:
Maybe with a strpos check on ; set a flag automatically to search for all words. So if ; is present in $expr search for all words.
Would that be a fast and easy fix?

Reply:
> Maybe with a strpos check on ; set a flag automatically to search for all words. So if ; is present in $expr search for all words.

That could work for your usecase, but what about "red shirt ; blue trousers ; black shoes" 😅 

I guess the very best would be
1. When we detect there are terms separated by `;`
2. We select each part separated by `;`
3. For each of these parts, we perform the regular intersect search : that means we look for red shirts, then we look for blue trousers, then we look for black shoes
4. Finally we merge all results together, so we return results that are red shirts OR blue trousers OR black shoes

It's a more complex though. Also such a complex behavior cannot be guessed, we will need
- to test it
- to document it

Reply:
At this point it's probably a good idea to ask @PrestaShop/prestashop-core-developers for their opinion

Reply:
"That could work for your usecase, but what about "red shirt ; blue trousers ; black shoes""
I think this can be fixed quite easy ..if we do an explode on ';' in $expr build a foreach. 
At the end tie everything together.

This would fix everything imho.

I'm going to test it

Reply:
@matks  . I tested it. If we explode the expressions on semicolon. Loop it in a for each it will give the expected result. In all ways

Reply:
Tested it with code the find() function of Prestashop 1.7.7.5. However it won't work with the code in 'Develop' branche. It will mess up the result later on in the code. Working on that. The current outcome: $eligibleProducts2Full (new var) has 2 product id's however only 1 product is returned as final result.

Reply:
PR updated!